### PR TITLE
Create unit tests for ParserUtil#parseIndex() #401

### PR DESCRIPTION
**Pull request by limmlingg**
Part of #401

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@limmlingg submitted v1 for review.

* [v1 1/1] [Add unit tests for ParserUtil\#parseIndex\(\) ](https://github.com/se-edu/addressbook-level4/pull/423/commits/884d2ca54e84c7aa0f981d9d817389ceb3f17b1e)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/ef06553753b5073c012ccac9ab1082c46b9d0f51...884d2ca54e84c7aa0f981d9d817389ceb3f17b1e))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/423/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 1]
```diff
@@ -22,11 +22,11 @@
 public class ParserUtil {
 
     /**
-     * Returns the specified index in the {@code command}.
+     * Returns the specified index in the {@code command}. Spaces will be trimmed.
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: `leading and trailing spaces will be trimmed`
Also, not {@code command} but {@code index}

Perhaps we can change the phrasing as well, now sounds quite weird. Maybe something like: `Parses {@code index} into an integer...`

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 2]
```diff
@@ -22,11 +22,11 @@
...
      * @throws IllegalValueException if the specified index is invalid (not positive unsigned integer).
      */
-    public static int parseIndex(String command) throws IllegalValueException {
-        String index = command.trim();
+    public static int parseIndex(String indexAsString) throws IllegalValueException {
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: can rename `indexAsString` to be `index`. then upon trimming, name it as `trimmedIndex`.

**Comment by CanIHasReview** [comment]
# v2

@limmlingg submitted v2 for review.

* [v2 1/1] [Add unit tests for ParserUtil\#parseIndex\(\) ](https://github.com/se-edu/addressbook-level4/pull/423/commits/b36c60e3115b3018ec28b509f44ca2bb16550f9f)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/ef06553753b5073c012ccac9ab1082c46b9d0f51...b36c60e3115b3018ec28b509f44ca2bb16550f9f)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/ef52bbb275eec8061d393ab59a5447cea29a9f47~1...ef52bbb275eec8061d393ab59a5447cea29a9f47))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/423/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/test/java/seedu/address/logic/parser/ParserUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        ParserUtil.parseIndex("10 a");
+    }
+
+    @Test
+    public void parseIndex_validInput_success() throws Exception {
+        // No whitespaces
+        assertEquals(1, ParserUtil.parseIndex("1"));
+
+        // Leading and trailing whitespaces
+        assertEquals(1, ParserUtil.parseIndex("  1  "));
```
Comment (eugenepeh) [inline src/test/java/seedu/address/logic/parser/ParserUtilTest.java]: The parseIndex test did not check for int out of range

`assertEquals(3423423423423432, ParserUtil.parseIndex("  3423423423423432  "));`
>java.lang.Error: Unresolved compilation problem: 
	The literal 3423423423423432 of type int is out of range

Comment (eugenepeh) [inline src/test/java/seedu/address/logic/parser/ParserUtilTest.java]: Running the command `delete 12312432535323212321` results in
>Caused by: java.lang.NumberFormatException: For input string: "12312432535323212321"

**Comment by yamgent** [comment]
> `assertEquals(3423423423423432, ParserUtil.parseIndex(" 3423423423423432 "));`
> > java.lang.Error: Unresolved compilation problem:
> > The literal 3423423423423432 of type int is out of range

This is actually a compilation error...`3423423423423432` is too big for an `int`. The correct way would be to expect an Exception to be thrown:

```
thrown.expect(IllegalValueException.class);
ParserUtil.parseIndex(" 3423423423423432 ");
```

> Running the command `delete 12312432535323212321` results in
> > Caused by: java.lang.NumberFormatException: For input string: "12312432535323212321"

Agree that we should add a test for checking out of range.

**Comment by yamgent** [comment]
Perhaps we can just inline the logic of `StringUtil#parseInteger()`, since it doesn't seem to be used by anyone else. We can fix these integer issues in `ParserUtil#parseIndex()` itself.

**Comment by Zhiyuan-Amos** [comment]
Yup to add on to our team lead's point, you can bring the StringUtil tests into ParserUtil. Then we have to make sure that the tests are complete :>

**Comment by damithc** [comment]
> Perhaps we can just inline the logic of StringUtil#parseInteger(), since it doesn't seem to be used by anyone else. We can fix these integer issues in ParserUtil#parseIndex() itself.

Move the method to `ParserUtil` as a private method? We shouldn't inline just because it's used only once. If something can be stated at a higher abstraction, then it is better to extract it as a method.

**Comment by limmlingg** [comment]
I don't know if that's part of the scope for this PR. Maybe we should address this in another issue?

**Comment by Zhiyuan-Amos** [comment]
@yamgent I'm thinking that we can create the test for large numbers in this PR, because it's testing the parseIndex method. For inlining... maybe a separate PR?

**Comment by yamgent** [comment]
Yes I think we can do the inline in a separate PR. For now we can just add the lerge number test in `ParserUtil#parseIndex()`?

**Comment by Zhiyuan-Amos** [comment]
yup sounds good. @limmlingg take note :P

**Comment by CanIHasReview** [comment]
# v3

@limmlingg submitted v3 for review.

* [v3 1/1] [Add unit tests for ParserUtil\#parseIndex\(\) ](https://github.com/se-edu/addressbook-level4/pull/423/commits/2d64411a222ba31584a8d9e637d649b6cd2f7cbc)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cc22ed642b34f15798facec1bddd597aa8c0c395...2d64411a222ba31584a8d9e637d649b6cd2f7cbc)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/81ccb2abd9c0b864ec86c98cced1b8b174b79a54~1...81ccb2abd9c0b864ec86c98cced1b8b174b79a54))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/423/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by eugenepeh** [comment]
Adding the test itself does not solve the issue that 
>Running the command `delete 12312432535323212321` results in
>>Caused by: java.lang.NumberFormatException: For input string: "12312432535323212321"

Perhaps this issue should be addressed in a separate PR first before merging this one?

**Comment by yamgent** [comment]
> Perhaps this issue should be addressed in a separate PR first before merging this one?

Ok, we can fix `StringUtil#isUnsignedInteger(String)` in a separate PR then. I have no objections for which PR this problem gets fixed.

**Comment by Zhiyuan-Amos** [comment]
Yup, the test should remain because it is documenting the bug / the intended effect due to our buggy code. We can fix `StringUtil#isUnsignedInteger(String)` in a separate PR.

**Comment by yamgent** [comment]
@eugenepeh Perhaps we can still get this merged first because it is an inherent problem with `StringUtil#isUnsignedInteger(String)` method, rather than `ParserUtil#parseIndex(String)`? Arguably `StringUtil#isUnsignedInteger(String)` should have been responsible for checking whether the number would even fit an `int`.

For another issue, @Zhiyuan-Amos pointed out a flaw in the name `isUnsignedInteger()`, because 0 is not accepted.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 4]
```diff
@@ -22,15 +22,15 @@
 public class ParserUtil {
 
     /**
-     * Returns the specified index in the {@code command}.
+     * Parses {@code index} into an integer and returns it. Leading and trailing whitespaces will be trimmed.
      * @throws IllegalValueException if the specified index is invalid (not positive unsigned integer).
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: maybe `non-zero unsigned integer` is clearer. `positive unsigned integer` is technically correct, but not as clear as `non-zero imo` :<

If you fix this, rmb to fix the exception string below as well :)

**Comment by eugenepeh** [comment]
@yamgent yeah that's fine, just pointing that out cause the test case might need to change (if catch and throw IllegalValueException) upon fixing that issue. Will open an issue for that.

 Its good to merge, just need to fix the comment issue Amos pointed out.

**Comment by CanIHasReview** [comment]
# v4

@limmlingg submitted v4 for review.

* [v4 1/1] [Add unit tests for ParserUtil\#parseIndex\(\) ](https://github.com/se-edu/addressbook-level4/pull/423/commits/c33f5f212876b7442f696100aa63480db248cf03)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/6af393367c998910784881146ffa049725bc31bd...c33f5f212876b7442f696100aa63480db248cf03)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/28eb19235a9e892f5e1a51cc184a06845332434a~1...28eb19235a9e892f5e1a51cc184a06845332434a))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/423/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>
